ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -272,9 +282,57 @@ func (r *RayServiceReconciler) calculateStatus(ctx context.Context, rayServiceIn
 		return errstd.New("numServeEndpoints exceeds math.MaxInt32")
 	}
 	rayServiceInstance.Status.NumServeEndpoints = int32(numServeEndpoints) //nolint:gosec // This is a false positive from gosec. See https://github.com/securego/gosec/issues/1212 for more details.
+	calculateConditions(rayServiceInstance)
 	return nil
 }
 
+func initConditions(rayServiceInstance *rayv1.RayService) (changed bool) {
+	changed = false
+	if rayServiceInstance.Status.Conditions == nil {
+		rayServiceInstance.Status.Conditions = []metav1.Condition{}
+		changed = true
+	}
+	if len(rayServiceInstance.Status.Conditions) == 0 {
+		message := "RayService is initializing"
+		setCondition(rayServiceInstance, rayv1.RayServiceReady, metav1.ConditionFalse, rayv1.RayServiceInitializing, message)
+		setCondition(rayServiceInstance, rayv1.UpgradeInProgress, metav1.ConditionFalse, rayv1.RayServiceInitializing, message)
+		changed = true
+	}
+	return changed
+}
+
+func calculateConditions(rayServiceInstance *rayv1.RayService) {
+	if rayServiceInstance.Status.NumServeEndpoints > 0 {
+		setCondition(rayServiceInstance, rayv1.RayServiceReady, metav1.ConditionTrue, rayv1.NonZeroServeEndpoints, "Number of serve endpoints is greater than 0")
+	} else if meta.IsStatusConditionTrue(rayServiceInstance.Status.Conditions, string(rayv1.RayServiceReady)) {
+		setCondition(rayServiceInstance, rayv1.RayServiceReady, metav1.ConditionFalse, rayv1.ZeroServeEndpoints, "Number of serve endpoints dropped to 0")
+	}
+
+	activeClusterName := rayServiceInstance.Status.ActiveServiceStatus.RayClusterName
+	pendingClusterName := rayServiceInstance.Status.PendingServiceStatus.RayClusterName
+	if activeClusterName != "" && pendingClusterName != "" {
+		setCondition(rayServiceInstance, rayv1.UpgradeInProgress, metav1.ConditionTrue, rayv1.BothActivePendingClustersExist, "Both active and pending Ray clusters exist")
+	} else if activeClusterName != "" {
+		setCondition(rayServiceInstance, rayv1.UpgradeInProgress, metav1.ConditionFalse, rayv1.NoPendingCluster, "Active Ray cluster exists and no pending Ray cluster")
+	} else {
+		cond := meta.FindStatusCondition(rayServiceInstance.Status.Conditions, string(rayv1.UpgradeInProgress))
+		if cond == nil || cond.Reason != string(rayv1.RayServiceInitializing) {
+			setCondition(rayServiceInstance, rayv1.UpgradeInProgress, metav1.ConditionUnknown, rayv1.NoActiveCluster, "No active Ray cluster exists, and the RayService is not initializing. Please open a GitHub issue in the KubeRay repository.")
+		}
+	}
+}
+
+func setCondition(rayServiceInstance *rayv1.RayService, conditionType rayv1.RayServiceConditionType, status metav1.ConditionStatus, reason rayv1.RayServiceConditionReason, message string) {
+	condition := metav1.Condition{
+		Type:               string(conditionType),
+		Status:             status,
+		Reason:             string(reason),
+		Message:            message,
+		ObservedGeneration: rayServiceInstance.Status.ObservedGeneration,
```

Comment:
The `ObservedGeneration` is not well maintained. Maybe we can avoid adding `ObservedGeneration` for now.